### PR TITLE
Implementation of issue # 310 - Custom title glitch for bar button items.

### PR DESCRIPTION
**Pull request by umerkhan-apple**
https://github.com/ResearchKit/ResearchKit/issues/310

### CODE REVIEW

**Comment by YuanZhu-apple** [comment]
@oliverschaefer Can you review this one? Explanation is here: https://github.com/ResearchKit/ResearchKit/issues/310#issuecomment-120160523

**Comment by oliverschaefer** [comment]
I think you need to call setupButtons after the delegate callback in viewWillAppear, not before.

I've noticed a second follow-up issue while reviewing: the learn more button title won't ever change, although the corresponding property is set correctly. Can you please check this?

**Comment by oliverschaefer** [comment]
The custom learnMoreButtonTitle gets considered once you move the stepDidChange call from viewDidLoad to viewWillAppear. Might this have any unwanted side effects?

**Comment by umerkhan-apple** [comment]
@oliverschaefer Moving the `[self setupButtons]` after the delegate callback in `viewWillAppear:` causes the back button to not attach the action specified by the custom button. Though, adding `[self stepDidChange]` after the delegate callback resolves the issue of `learnMoreButtonTitle` not updating. I have not encountered any unwanted side effects as of yet. Im pushing the changes, please review.

**Comment by oliverschaefer** [comment]
@umerkhan-apple As long as `[self setupButtons]` is called before the delegate callback, I still encounter the glitch while navigating?!

**Comment by umerkhan-apple** [comment]
@oliverschaefer I am not able to recreate the glitch after this fix. Could you elaborate as to exactly what you are encountering? 

As for the `learnMoreButtonTitle`, the button had to be reset after the the title property is set. I have updated the PR to reflect these changes. As a result, we no longer need `[self stepDidChange]' after the delegate callback.

**Comment by oliverschaefer** [comment]
@umerkhan-apple Okay, I finally figured out what was going on. My delegate implementation of `stepViewControllerWillAppear` tried to change the title of the `cancelButtonItem` via
`stepViewController.cancelButtonItem?.title = "new title"`. That led to a noticeable glitch when navigating to another stepViewController, even with your fix in place. After re-reading the documentation of the `cancelButtonItem` property, I now use `stepViewController.cancelButtonItem = UIBarButtonItem(title: "new title", style: .Plain, target: stepViewController.cancelButtonItem?.target, action: stepViewController.cancelButtonItem?.action ?? "")`. That way the glitch is gone, so it should be save to merge the PR now.

@umerkhan-apple @YuanZhu-apple @jwe-apple 
However, it is a little confusing to modify titles by changing dedicated properties on the one hand (such as `learnMoreButtonTitle`) and by setting custom button items on the other hand (such as `cancelButtonItem`), but that leads to longer discussion (is it really advisable to provide your own task cancelation logic via custom targets/actions?)

**Comment by umerkhan-apple** [comment]
@oliverschaefer It is really not advisable to provide your own task cancelation logic. The reason for setting the cancelButtonItem with a custom button is to allow the user to pass a nil value. ``When the value of the property is `nil`, the Cancel button is not displayed;`` On the other hand, `learnMoreButton` has a delegate callback to allow for customization.

**Comment by oliverschaefer** [comment]
Okay, I see.

File: ResearchKit/Common/ORKStepViewController.m [hunk 1]
```diff
@@ -140,12 +139,13 @@ - (void)stepDidChange {
 - (void)viewWillAppear:(BOOL)animated {
     [super viewWillAppear:animated];
     ORK_Log_Debug(@"%@", self);
+    
     [self setupButtons];
```
Comment (YuanZhu-apple) [inline ResearchKit/Common/ORKStepViewController.m]: Can you add some documentation for why this is required at here? In case we forget next time:)

Comment (umerkhan-apple) [inline ResearchKit/Common/ORKStepViewController.m]: @YuanZhu-apple Good call. PR coming up!

**Comment by YuanZhu-apple** [comment]
r=me.

**Pull request merged**